databricks/koalas/sql.py

```diff
@@ -0,0 +1,286 @@
+    precedence as globals.
+
+    Parameters
+    ----------
+    query : str
+        the SQL query
+
+    globals : the dictionary of global variables, if explicitly set by the user
+
+    locals : the dictionary of local variables, if explicitly set by the user
+
+    kwargs : other variables that the user may want to set manually.
+
+    Returns
+    -------
+    DataFrame
+
+    Examples
+    --------
+
+    Calling a built-in SQL function.
+
+    >>> ks.sql("select * from range(10) where id > 7")
+       id
+    0   8
+    1   9
+
+    A query can also reference a local variable or parameter by wrapping them in curly braces:
+
+    >>> bound1 = 7
+    >>> ks.sql("select * from range(10) where id > {bound1} and id < {bound2}", bound2=9)
+       id
+    0   8
+
+    You can also wrap a DataFrame with curly braces to query it directly. Note that when you do
+    that, the indexes, if any, automatically become top level columns.
+
+    >>> mydf = ks.range(10)
+    >>> x = range(4)
+    >>> ks.sql("SELECT * from {mydf} WHERE id IN {x}")
+       id
+    0   0
+    1   1
+    2   2
+    3   3
+
+    Queries can also be arbitrarily nested in functions:
+
+    >>> def statement():
+    ...     mydf2 = ks.DataFrame({"x": range(2)})
+    ...     return sql("SELECT * from {mydf2}")
+    >>> statement()
+       __index_level_0__  x
+    0                  0  0
+    1                  1  1
+
+    Mixing Koalas and pandas DataFrames in a join operation. Note that the index is
+    dropped.
+
+    >>> sql('''
+    ...   SELECT m1.a, m2.b
+    ...   FROM {table1} m1 INNER JOIN {table2} m2
+    ...   ON m1.key = m2.key
+    ...   ORDER BY m1.a, m2.b''',
+    ...   table1=ks.DataFrame({"a": [1,2], "key": ["a", "b"]}),
+    ...   table2=pd.DataFrame({"b": [3,4,5], "key": ["a", "b", "b"]}))
+       a  b
+    0  1  3
+    1  2  4
+    2  2  5
+
+    """
+    if globals is None:
+        globals = _get_ipython_scope()
+    _globals = builtin_globals() if globals is None else dict(globals)
+    _locals = builtin_locals() if locals is None else dict(locals)
+    # The default choice is the globals
+    _dict = dict(_globals)
+    # The vars:
+    _scope = _get_local_scope()
+    _dict.update(_scope)
+    # Then the locals
+    _dict.update(_locals)
+    # Highest order of precedence is the locals
+    _dict.update(kwargs)
+    return SQLProcessor(_dict, query, default_session()).execute()
+
+
+def _get_local_scope():
+    # Get 2 scopes above (_get_local_scope -> sql -> ...) to capture the vars there.
+    try:
+        return inspect.stack()[2][0].f_locals
+    except Exception as e:
+        return {}
+
+
+def _get_ipython_scope():
+    """
+    Tries to extract the dictionary of variables if the program is running
+    in an IPython notebook environment.
+    """
+    try:
+        from IPython import get_ipython
+        shell = get_ipython()
+        return shell.user_ns
+    except Exception as e:
```

Comment:
Same as above.

Reply:
@thunterdb were you expecting a more specific exception?

Reply:
@rxin this is also a catch all for:
 - import errors
 - ipython not initialized or not running
Because it tries to catpure the variables in the scope without checking really the intent of the user (are we running in jupyter?), it is rather heavy-handed.

There is probably a better way by looking at the internals of how python implements string interpolation (and can deal with expressions), but I could not find something simple. I believe that the current implementation is good enough, but let me know if you have some thoughts on this.